File: node/src/sensor/gas.ts
import { Sensor, SensorType, SensorCategory } from './base.js';
import { defineSensor, SensorDomain } from './meta.js';

import type { Observable } from '../observable/index.js';
import type { PropertyChangeOf, SensorLike, SensorOptions } from './base.js';

/**
 * Property names of a gas sensor.
 *
 * @internal
 */
export enum GasProperty {
  /** Whether gas is detected. */
  Detected = 'detected',
}

/**
 * Property values of a gas sensor.
 *
 * @internal
 */
export interface GasSensorProperties {
  [GasProperty.Detected]: boolean;
}

/** Read-only proxy interface for a gas sensor. */
export interface GasSensorLike extends SensorLike {
  readonly type: SensorType.Gas;
  readonly onPropertyChanged: Observable<PropertyChangeOf<GasSensorProperties>>;

  getValue(property: GasProperty.Detected): boolean | undefined;
  getValue(property: string): unknown;
}

/** Gas detector sensor. */
export class GasSensor<TStorage extends object = Record<string, any>> extends Sensor<GasSensorProperties, TStorage> {
  readonly type = SensorType.Gas;
  readonly category = SensorCategory.Sensor;

  constructor(name = 'Gas Sensor', options?: SensorOptions) {
    super(name, options);

    this._writeState({ [GasProperty.Detected]: false });
  }

  get detected(): boolean {
    return this.props.detected;
  }

  /**
   * Report gas detection state.
   *
   * @param value - True when gas is currently detected.
   *
   * @example
   * ```ts
   * gas.setDetected(true);
   * ```
   */
  setDetected(value: boolean): void {
    this._writeState({ [GasProperty.Detected]: value });
  }

  /**
   * Read-only sensor: external writes are ignored.
   *
   * @internal
   */
  updateValue(_property: string, _value: unknown): void {}
}

/** Registry metadata for {@link GasSensor}. */
export const gasMeta = defineSensor({
  type: SensorType.Gas,
  category: SensorCategory.Sensor,
  assignmentKey: 'gas',
  multiProvider: true,
  isDetectionType: false,
  properties: {
    [GasProperty.Detected]: { type: 'boolean', writable: true },
  },
  shortcutable: true,
  cascadeTrigger: { property: GasProperty.Detected, value: true, sustained: true },
  virtual: { properties: { [GasProperty.Detected]: false } },
  semantics: {
    domain: SensorDomain.Binary,
    stateProperty: GasProperty.Detected,
    commandProperty: GasProperty.Detected,
    deviceClass: 'gas',
  },
});

File: python/camera_ui_sdk/sensor/gas.py
from __future__ import annotations

from collections.abc import Mapping
from enum import StrEnum
from typing import Any, Generic, Literal, Protocol, overload, runtime_checkable

from typing_extensions import TypedDict, TypeVar

from ..observable import Observable
from .base import Sensor, SensorCategory, SensorLike, SensorType


class GasProperty(StrEnum):
    """Property names of a gas sensor."""

    Detected = "detected"
    """Whether gas is detected."""


class GasSensorProperties(TypedDict):
    """Property values of a gas sensor."""

    detected: bool


class GasPropertyChangeData(TypedDict):
    """Property change payload emitted on GasSensorLike.onPropertyChanged."""

    property: str
    """Name of the changed property, a GasProperty value."""
    value: bool
    """New value of the property."""


TStorage = TypeVar("TStorage", bound=Mapping[str, Any], default=dict[str, Any])


@runtime_checkable
class GasSensorLike(SensorLike, Protocol):
    """Read-only proxy interface for a gas sensor."""

    @property
    def type(self) -> SensorType:
        return SensorType.Gas

    @property
    def onPropertyChanged(self) -> Observable[GasPropertyChangeData]: ...

    @overload
    def getValue(self, property: Literal[GasProperty.Detected]) -> bool | None: ...
    @overload
    def getValue(self, property: str) -> object | None: ...


class GasSensor(Sensor[GasSensorProperties, TStorage, str], Generic[TStorage]):
    """Gas detector sensor."""

    _requires_frames = False

    def __init__(self, name: str = "Gas Sensor", *, native_id: str | None = None) -> None:
        super().__init__(name, native_id=native_id)
        self._write_state({GasProperty.Detected.value: False})

    @property
    def type(self) -> SensorType:
        return SensorType.Gas

    @property
    def category(self) -> SensorCategory:
        return SensorCategory.Sensor

    @property
    def detected(self) -> bool:
        return bool(self.props.detected)

    def setDetected(self, value: bool) -> None:
        """Report gas detection state.

        Args:
            value: True when gas is currently detected.

        Example:
            ```python
            gas.setDetected(True)
            ```
        """
        self._write_state({GasProperty.Detected.value: value})

    async def updateValue(self, property: str, value: Any) -> None:
        """Read-only sensor: external writes are ignored."""
